openssl-sys/src/lib.rs

```diff
@@ -13,162 +13,174 @@
 #![recursion_limit = "128"] // configure fixed limit across all rust versions
 
 extern crate libc;
+pub use libc::*;
+
+#[cfg(boringssl)]
+extern crate bssl_sys;
+#[cfg(boringssl)]
+pub use bssl_sys::*;
+
+#[cfg(openssl)]
+#[path = "."]
+mod openssl {
+    use libc::*;
+
+    #[cfg(feature = "bindgen")]
+    include!(concat!(env!("OUT_DIR"), "/bindgen.rs"));
+
+    pub use self::aes::*;
+    pub use self::asn1::*;
+    pub use self::bio::*;
+    pub use self::bn::*;
+    pub use self::cms::*;
+    pub use self::crypto::*;
+    pub use self::dtls1::*;
+    pub use self::ec::*;
+    pub use self::err::*;
+    pub use self::evp::*;
+    #[cfg(not(feature = "bindgen"))]
+    pub use self::handwritten::*;
+    pub use self::obj_mac::*;
+    pub use self::ocsp::*;
+    pub use self::pem::*;
+    pub use self::pkcs7::*;
+    pub use self::rsa::*;
+    pub use self::sha::*;
+    pub use self::srtp::*;
+    pub use self::ssl::*;
+    pub use self::ssl3::*;
+    pub use self::tls1::*;
+    pub use self::types::*;
+    pub use self::x509::*;
+    pub use self::x509_vfy::*;
+    pub use self::x509v3::*;
+
+    #[macro_use]
+    mod macros;
+
+    mod aes;
+    mod asn1;
+    mod bio;
+    mod bn;
+    mod cms;
+    mod crypto;
+    mod dtls1;
+    mod ec;
+    mod err;
+    mod evp;
+    #[cfg(not(feature = "bindgen"))]
+    mod handwritten;
+    mod obj_mac;
+    mod ocsp;
+    mod pem;
+    mod pkcs7;
+    mod rsa;
+    mod sha;
+    mod srtp;
+    mod ssl;
+    mod ssl3;
+    mod tls1;
+    mod types;
+    mod x509;
+    mod x509_vfy;
+    mod x509v3;
+
+    // FIXME remove
+    pub type PasswordCallback = unsafe extern "C" fn(
+        buf: *mut c_char,
+        size: c_int,
+        rwflag: c_int,
+        user_data: *mut c_void,
+    ) -> c_int;
+
+    #[cfg(ossl110)]
+    pub fn init() {
+        use std::ptr;
+        use std::sync::Once;
+
+        // explicitly initialize to work around https://github.com/openssl/openssl/issues/3505
+        static INIT: Once = Once::new();
+
+        #[cfg(not(ossl111b))]
+        let init_options = OPENSSL_INIT_LOAD_SSL_STRINGS;
+        #[cfg(ossl111b)]
+        let init_options = OPENSSL_INIT_LOAD_SSL_STRINGS | OPENSSL_INIT_NO_ATEXIT;
```

Comment:
This fixes some crashes I was experiencing in diesel using libpq, which uses openssl. The atexit handler was racing the r2d2 connection pool worker threads when main() exits, and that caused a seg fault.

See: https://github.com/sfackler/r2d2/issues/137

Reply:
👍🏼  Reproduced using MySql also